### ISSUE TITLE
Passenger 5.0.18 on CentOS-7 does not automatically scales the number of processes based on traffic for nodejs

### ISSUE THREAD

**Issue by xyztest** [phusion/passenger#1618]
I use passenger to start a testing server, copied from the app.js from the passenger nodejs demo. The following are the output when first starting the service.

~~~
[dtian@drtest01 passenger-nodejs-connect-demo]$ passenger start --app-type node --startup-file server.js
=============== Phusion Passenger Standalone web server started ===============
PID file: /home/dtian/passenger-nodejs-connect-demo/passenger.3000.pid
Log file: /home/dtian/passenger-nodejs-connect-demo/passenger.3000.log
Environment: development
Accessible via: http://0.0.0.0:3000/

You can stop Phusion Passenger Standalone by pressing Ctrl-C.
Problems? Check https://www.phusionpassenger.com/library/admin/standalone/troubleshooting/
===============================================================================
App 9136 stdout: 
App 9136 stdout: Example app running inside Passenger.
App 9136 stdout: {"name":"serverlog","hostname":"drtest01.plaxo.com","pid":9136,"level":30,"msg":"Responded to http request at /","time":"2015-09-22T22:40:43.258Z","v":0}
App 9136 stdout: {"name":"serverlog","hostname":"drtest01.plaxo.com","pid":9136,"level":50,"msg":"Test error msg at /","time":"2015-09-22T22:40:43.260Z","v":0}
~~~

Then I go to another host to run benchmark testing by using the ab tool. It has the following output.

~~~
[dtian@ci05 ~]$ ab -n 50000  -c 10 http://172.20.6.13:3000/
This is ApacheBench, Version 2.3 <$Revision: 1430300 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 172.20.6.13 (be patient)
Completed 5000 requests
Completed 10000 requests
Completed 15000 requests
Completed 20000 requests
Completed 25000 requests
Completed 30000 requests
Completed 35000 requests
Completed 40000 requests
Completed 45000 requests
Completed 50000 requests
Finished 50000 requests


Server Software:        nginx/1.8.0
Server Hostname:        172.20.6.13
Server Port:            3000

Document Path:          /
Document Length:        39 bytes

Concurrency Level:      10
Time taken for tests:   45.722 seconds
Complete requests:      50000
Failed requests:        0
Write errors:           0
Total transferred:      10900000 bytes
HTML transferred:       1950000 bytes
Requests per second:    1093.57 [#/sec] (mean)
Time per request:       9.144 [ms] (mean)
Time per request:       0.914 [ms] (mean, across all concurrent requests)
Transfer rate:          232.81 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.2      0       4
Processing:     2    9   2.1      8      56
Waiting:        2    9   2.1      8      56
Total:          3    9   2.1      9      57

Percentage of the requests served within a certain time (ms)
  50%      9
  66%     10
  75%     10
  80%     10
  90%     11
  95%     12
  98%     16
  99%     18
 100%     57 (longest request)

When the ab command execution is complete, there passenger-status shows only one server process handling the queries. It seems passenger does not auto scale the number of processes to handle heavy traffic.

[dtian@drtest01 passenger-nodejs-connect-demo]$ passenger-status
Version : 5.0.18
Date    : 2015-09-22 15:42:50 -0700
Instance: pVJeLwkd (nginx/1.8.0 Phusion_Passenger/5.0.18)

----------- General information -----------
Max pool size : 6
App groups    : 1
Processes     : 1
Requests in top-level queue : 0

----------- Application groups -----------
/home/dtian/passenger-nodejs-connect-demo/public (development):
  App root: /home/dtian/passenger-nodejs-connect-demo
  Requests in queue: 0
  * PID: 9136    Sessions: 0       Processed: 50001   Uptime: 2m 7s
    CPU: 37%     Memory  : 48M     Last used: 26s ago


Indeed, the ab test output shows the response numbers for service running in passenger is worse than the response numbers for native node service without running passenger.

Here is the command to start native node service: node server.js

Here are the output of ab testing results, it shows shorter response time.


[dtian@ci05 ~]$ ab -n 50000  -c 10 http://172.20.6.13:3000/
This is ApacheBench, Version 2.3 <$Revision: 1430300 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 172.20.6.13 (be patient)
Completed 5000 requests
Completed 10000 requests
Completed 15000 requests
Completed 20000 requests
Completed 25000 requests
Completed 30000 requests
Completed 35000 requests
Completed 40000 requests
Completed 45000 requests
Completed 50000 requests
Finished 50000 requests


Server Software:        
Server Hostname:        172.20.6.13
Server Port:            3000

Document Path:          /
Document Length:        39 bytes

Concurrency Level:      10
Time taken for tests:   23.824 seconds
Complete requests:      50000
Failed requests:        0
Write errors:           0
Total transferred:      5700000 bytes
HTML transferred:       1950000 bytes
Requests per second:    2098.76 [#/sec] (mean)
Time per request:       4.765 [ms] (mean)
Time per request:       0.476 [ms] (mean, across all concurrent requests)
Transfer rate:          233.65 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    1   0.3      0       6
Processing:     2    4   0.8      4      26
Waiting:        2    4   0.8      4      26
Total:          3    5   0.9      5      26
ERROR: The median and mean for the initial connection time are more than twice the standard
       deviation apart. These results are NOT reliable.

Percentage of the requests served within a certain time (ms)
  50%      5
  66%      5
  75%      5
  80%      5
  90%      5
  95%      6
  98%      7
  99%      9
 100%     26 (longest request)
~~~

**Comment by FooBarWidget** [comment]
This is normal. Node.js apps are evented and are able to handle any arbitrary number of connections, so Passenger does not really know when to automatically spawn a new Node.js app process. You should tell Passenger how many processes it should spawn using `--min-instances`.

**Comment by majgis** [comment]
Are there any conditions that will cause a node.js app to auto-scale?

**Comment by OnixGH** [comment]
@majgis not at present, but we are actively looking into this.

**Comment by majgis** [comment]
It seems like this issue should be reopened or the documentation updated. There are multiple places in the documentation that state that auto-scaling works for node.js applications.

**Comment by OnixGH** [comment]
FYI: the documentation was already addressed via https://github.com/phusion/passenger_library/issues/9.

**Comment by FooBarWidget** [comment]
We will have a solution soon. We are testing a solution with an enterprise customer who is also on Node.js.

**Comment by majgis** [comment]
Good information, thank you for the update.

**Comment by FooBarWidget** [comment]
Implemented in 5.0.22: https://www.phusionpassenger.com/library/config/nginx/dynamic_scaling_vs_fixed_app_processes/nodejs/
